jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -352,117 +353,259 @@ public void abort(Throwable failure)
+                        // An idle timeout during these transient states is ignored.
+                    case FAILED:
+                        // Already failed, ignore the idle timeout.
+                    {
+                        callback = null;
+                        timeout = false;
+                        break;
+                    }
+                    default:
+                    {
+                        // Should not happen, but just in case.
+                        _state = State.FAILED;
+                        _failure = new IllegalStateException("Invalid transport state: " + _state, failure);
+                        callback = _callback;
+                        if (callback == null)
+                            callback = Callback.NOOP;
+                        timeout = true;
+                        failure = _failure;
+                        break;
+                    }
                 }
             }
+            idleTimeout(callback, timeout, failure);
+            return timeout;
+        }
+
+        private void succeed(Callback callback, boolean commit)
+        {
             if (LOG.isDebugEnabled())
-                LOG.debug(String.format("HTTP2 Response #%d/%h idle timeout %s", stream.getId(), stream.getSession(), result ? "expired" : "ignored"), failure);
-            if (result)
+                LOG.debug("HTTP2 Response #{}/{} {} success",
+                    stream.getId(), Integer.toHexString(stream.getSession().hashCode()),
+                    commit ? "commit" : "flush");
+            callback.succeeded();
+        }
+
+        private void fail(Callback callback, boolean commit, Throwable failure)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("HTTP2 Response #{}/{} {} failure",
+                    stream.getId(), Integer.toHexString(stream.getSession().hashCode()),
+                    commit ? "commit" : "flush",
+                    failure);
+            if (callback != null)
                 callback.failed(failure);
-            return result;
         }
 
-        @Override
-        public InvocationType getInvocationType()
+        private void idleTimeout(Callback callback, boolean timeout, Throwable failure)
         {
-            Callback callback;
-            synchronized (this)
-            {
-                callback = this.callback;
-            }
-            return callback != null ? callback.getInvocationType() : Callback.super.getInvocationType();
+            if (LOG.isDebugEnabled())
+                LOG.debug("HTTP2 Response #{}/{} idle timeout {}",
+                    stream.getId(), Integer.toHexString(stream.getSession().hashCode()),
+                    timeout ? "expired" : "ignored",
+                    failure);
+            if (timeout)
+                callback.failed(failure);
         }
     }
 
     private enum State
     {
-        IDLE, WRITING, FAILED, TIMEOUT
+        IDLE, PRE_SEND, POST_SEND, SUCCEED, FAIL, FAILED
```

Comment:
These states and their transitions should be javadoc'ed as it's not easy to grok what the FSM looks like by reading the code.

Reply:
Done.